RFC-0009-jdbc-join-push-down.md

```diff
@@ -0,0 +1,607 @@
+
+Examples to explain presto change an inner join to another Join operation : 
+
+Suppose we have a query like this: 
+
+`Select * from table a join table b on a.col1 = b.col2 and a.col1 = 5;`
+
+Presto will change this from an inner join to two different select statements like this: 
+
+`Select * from table a where a.col1 = 5;`
+
+`Select * from table b where b.col2 = 5;`
+
+Then it does a cross join with these two results. We will not do pushdown in this case.
+
+3) Join criteria (joining column) should be of Datatypes and operators that support join pushdown. 
+
+| No | DataType support join pushdown                   | Operations                                           |
+|----|-------------------------------------|-------------------------------------------------------------------|
+| 1  | TinyINT        | `=, <, >, <=, >=, !=, <>`                    |
+| 2  | SmallINT       | `=, <, >, <=, >=, !=, <>`                    |   
+| 3  | Integer        | `=, <, >, <=, >=, !=, <>`                    |
+| 4  | BigINT         | `=, <, >, <=, >=, !=, <>`                    |
+| 5  | Boolean        | `=, !=, <>`                                  |   
+| 6  | Integer        | `=, <, >, <=, >=, !=, <>`                    |
+| 7  | Real           | `=, <, >, <=, >=, !=, <>`                    |   
+| 8  | Double         | `=, <, >, <=, >=, !=, <>`                    |
+| 9  | Decimal        | `=, <, >, <=, >=, !=, <>`                    |
+| 10 | Varchar        | `=, <, >, <=, >=, !=, <>`                    |
+| 11 | Char           | `=, <, >, <=, >=, !=, <>`                    |
+
+
+4) All tables from same connector will be grouped based on above specifications and pushed down to underlying datasource. 
+
+5) Enable presto Join pushdown capabilities by setting the session flag enable-join-query-pushdown = true.
+
+## Low level Design
+
+For performing Jdbc JoinPushdown we required below implementations :
+
+1. Create 2 Optimizers called GroupInnerJoinsByConnector and JdbcJoinPushdown
+
+For Join pushingdown we need to travers through the JoinNode. We are able to create a new optimizer (GroupInnerJoinsByConnector) which implements PlanOptimizer and another optimizer JdbcJoinPushdown which implements ConnectorPlanOptimizer.
+
+ Below is the overall process that we have in GroupInnerJoinsByConnector optimizer
+
+![Image 1](RFC-0009-jdbc-join-push-down/in-depth-design-image-1.png)
+
+After completing GroupInnerJoinsByConnector optimization we need to invoke predicate pushdown optimizer to recreate join criteria from the filter node of the JoinNode. The detailed implantation of GroupInnerJoinsByConnector optimizer is explained in below sessions. 
```

Comment:
After GroupInnerJoinsByConnector we should invoke JdbcJoinPushdown and not the predicate pushdown optimizer. Predicate pushdown optimizer may invoke after JdbcJoinPushdown optimizer

Reply:
fixed